includes/components/class-woocommerce.php

```diff
@@ -379,14 +379,38 @@ public function set_account_template( $path, $name ) {
 	}
 
 	/**
-	 * Redirects account page.
+	 * Page redirection settings.
 	 */
-	public function redirect_account_page() {
+	public function redirect_pages() {
+
+		// Redirect for account page
 		if ( ! is_user_logged_in() && is_account_page() ) {
 			wp_safe_redirect( hivepress()->router->get_return_url( 'user_login_page' ) );
 
 			exit;
 		}
+
+		// Redirect from WC product page to related listing page
+		if ( is_product() ) {
+
+			// Get parent ID
+			$parent_id = wp_get_post_parent_id();
+
+			if ( ! $parent_id ) {
+				return;
+			}
+
+			// Get listing URL
+			$listing_url = hivepress()->router->get_url( 'listing_view_page', [ 'listing_id' => $parent_id ] );
```

Comment:
We don't know for sure if the parent ID is the listing ID, third-party plugins may also set their post types as product parents for different purposes. checking if the post type (e.g. if it starts with "hp_") should be enough.

Reply:
You can also make the callback more simple by setting an empty $url at the beginning and assign it in different conditions, then checking if the $url is not empty at the end and calling wp_safe_redirect once.